keps/sig-node/20190604-cgroup-none.md

```diff
@@ -0,0 +1,174 @@
+
+## Summary
+
+Add "none" cgroup driver to kubelet so that kubelet (as well as CRI/OCI runtimes) can be executed without write permissions to cgroups.
+
+Docker [v19.03+](https://github.com/moby/moby/pull/38050), containerd/CRI [20190104+](https://github.com/containerd/cri/pull/970), and CRI-O [v1.12+](https://github.com/cri-o/cri-o/pull/1729) already support running without cgroup permissions. So we only need to update kubelet.
+
+Draft patch for kubelet is available in [Usernetes project](https://github.com/rootless-containers/usernetes): https://github.com/rootless-containers/usernetes/blob/2fd0ba0594673e0c5ea04a894191bfaa3b117377/src/patches/kubernetes/0003-kubelet-new-feature-gate-SupportNoneCgroupDriver.patch
+
+An old version of our patch has been also used by Rancher's [k3s](https://github.com/rancher/k3s/commit/e397cb4e5b017805588ff223a63f42bff1f42d3a).
+
+## Motivation
+
+The motivation of adding "none" cgroup driver is to allow running the entire Kubernetes components (`kubelet`, CRI, OCI, CNI, and all `kube-*`) as a non-root user (w/o write access to `/sys/fs/cgroup`) on the host. ("Rootless mode").
+
+The motivations of Rootless mode are:
+* To protect the host from potential container-breakout vulnerabilities
+* To allow users of shared machines (especially HPC) to run Kubernetes without the risk of breaking their colleagues' environments
+* Safe Kubernetes-on-Kubernetes
+
+Without cgroups, kubelet cannot support restricting and monitoring CPU/memory resources for pods, but pod processes can still spontaneously impose traditional `setrlimit` restrictions on themselves.
+
+See [FOSDEM 2019 talk "Rootless Kubernetes"](https://www.slideshare.net/AkihiroSuda/rootless-kubernetes) by [@AkihiroSuda](https://github.com/AkihiroSuda) and [@giuseppe](https://github.com/giuseppe) for the overview of Rootless mode.
+
+### Future: cgroup2
+
+In future, we will be able to safely delegate cgroup permission to unprivileged users by using [cgroup2 unified-mode](https://systemd.io/CGROUP_DELEGATION.html#delegation).
+So most users will no longer need to use the "none" cgroup driver.
+
+However, the migration to cgroup2 is likely to take a couple of years or even more, because [cgroup1 controllers and cgroup2 controllers cannot be used simultaneously](https://systemd.io/CGROUP_DELEGATION.html#three-different-tree-setups-).
+([Fedora 31](https://www.phoronix.com/scan.php?page=news_item&px=Fedora-31-Cgroups-V2-Default) is already planning to use cgroup2 by default, but anyway Kubernetes users will need to
+change the GRUB configuration to use cgroup1, as Kubelet and CRI runtimes will not be able to support cgroup2 by the releae of Fedora 31.)
+
+Also, even after the migration to cgroup2, containerized Kubernetes might not get support for cgroup2, because it is likely that [cgroup2 will only be supported via systemd](https://github.com/containers/libpod/issues/1429#issuecomment-465420444). So the "none" cgroup driver will still remain useful.
```

Comment:
I am not following this chain of thought.  I prefer we work towards enabling cgroup v2 and rationalize what a cgroup manager role is in a cgroup v2 host environment.  I know we have discussed enabling the kubelet to run in cgroupv1 tolerant or cgroup v2 mode (with maybe a reduced set of controllers).  In general, I am apprehensive about disabling all resource isolation versus working towards an end-state that enables or supports it.  I think most users who hear rootless kubernetes will assume it comes with no limitations.